Quotient/e2ee/cryptoutils.cpp

```diff
@@ -59,75 +60,90 @@ inline std::pair<int, bool> checkedSize(
+        checkedSize((ByteArray_).size() __VA_OPT__(, ) __VA_ARGS__);       \
+    if (ByteArray_##Clamped) {                                             \
+        qCCritical(E2EE).nospace()                                         \
+            << __func__ << ": " #ByteArray_ " is " << (ByteArray_).size()  \
+            << " bytes long, too much for OpenSSL and overall suspicious"; \
+        Q_ASSERT(!ByteArray_##Clamped); /* Always false */                 \
+        return SslPayloadTooLong;                                          \
+    }                                                                      \
+    // End of macro
+
+#define CALL_OPENSSL(Call_)                                                \
+    do {                                                                   \
+        if ((Call_) <= 0) {                                                \
+            qCWarning(E2EE) << __func__ << "failed to call OpenSSL API:"   \
+                            << ERR_error_string(ERR_get_error(), nullptr); \
+            return ERR_get_error();                                        \
+        }                                                                  \
+    } while (false) // End of macro
+// NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
 SslExpected<QByteArray> Quotient::pbkdf2HmacSha512(const QByteArray& password,
                                                    const QByteArray& salt,
                                                    int iterations, int keyLength)
 {
+    CLAMP_SIZE(passwordSize, password)
+    CLAMP_SIZE(saltSize, salt)
     auto output = zeroedByteArray(keyLength);
-    if (!PKCS5_PBKDF2_HMAC(password.data(), password.size(), reinterpret_cast<const unsigned char *>(salt.data()), salt.length(), iterations, EVP_sha512(), keyLength, reinterpret_cast<unsigned char *>(output.data()))) {
-        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
-        return ERR_get_error();
-    }
+    CALL_OPENSSL(
+        PKCS5_PBKDF2_HMAC(password.data(), passwordSize, asCBytes(salt).data(),
+                          saltSize, iterations, EVP_sha512(), keyLength,
+                          reinterpret_cast<unsigned char*>(output.data())));
     return output;
 }
 
 SslExpected<QByteArray> Quotient::aesCtr256Encrypt(const QByteArray& plaintext,
-                                                   const QByteArray& key,
-                                                   const QByteArray& iv)
+                                                   byte_view_t<AesKeySize> key,
+                                                   byte_view_t<AesBlockSize> iv)
 {
-    const auto [plaintextSize, clamped] =
-        checkedSize(plaintext.size(),
-                    std::numeric_limits<int>::max() - AesBlockSize);
-    Q_ASSERT(!clamped); // Normally the caller should check this
+    CLAMP_SIZE(plaintextSize, plaintext,
+               std::numeric_limits<int>::max() - iv.size())
```

Comment:
what's the reason for `std::numeric_limits<int>::max() - iv.size()`?

Reply:
Yeah, looks like I got carried away a bit.